File: advection/output_mpi.f90
! Daniel R. Reynolds
! SMU Mathematics
! Math 4370/6370
! 7 February 2015
!=================================================================


subroutine output_mpi(u, t, nxloc, nyloc, pid, pcoords, &
                      nx, ny, px, py, noutput)
  !===============================================================
  ! Description: 
  !    Writes current solution to disk in parallel, with a 
  !    separate file for each MPI process.
  !===============================================================
  ! inclusions
  implicit none

  ! declarations
  integer, intent(in) :: nxloc, nyloc, pid, pcoords(2)
  integer, intent(in) :: nx, ny, px, py, noutput
  real*8,  intent(in) :: t, u(nxloc,nyloc)
  integer :: i, j
  character*50 :: outname
  
  ! internals
  
  ! set output file name
  ! Note: we reserve the first set of digits for the MPI process (unused here)
  write(outname,'(5Hu_sol,f4.3,f4.3)') float(pid)/1000.0 , noutput/1000.0

  ! open file
  open(102,file=outname)

  ! write data set parameters for this process
  write(102,*) nxloc
  write(102,*) nyloc
  write(102,*) pcoords(1)
  write(102,*) pcoords(2)
  write(102,*) t

  ! output the solution values and close the data set
  do j=1,nyloc
     do i=1,nxloc
        write(102,*) u(i,j)
     end do
  end do
  close(102)

  ! now output a metadata file, containing general run information
  if (pid == 0) then
     open(103,file='u_sol.txt')
     write(103,*) nx
     write(103,*) ny
     write(103,*) px
     write(103,*) py
     write(103,*) noutput
     close(103)
  end if

  return

  ! end program
end subroutine output_mpi
!=================================================================

File: advection/output_mpi.c
/* Daniel R. Reynolds
   SMU Mathematics
   Math 4370 / 6370 */

/* Inclusions */
#include <stdio.h>
#include <string.h>
#include "advection_mpi.h"
#include "mpi.h"


/* Writes current solution to disk. */
void output(double* u, double t, int nx, int ny, int noutput, parallel_decomp *p2d) {

  /* declarations */
  char outname[100];
  FILE* FID;
  int i, j;

  /* set output file name */
  sprintf(outname, "u_sol.%03i.%03i", p2d->myid, noutput);

  /* open output file */
  FID = fopen(outname,"w");

  /* write data set parameters */
  fprintf(FID, "%i\n", p2d->nxloc);
  fprintf(FID, "%i\n", p2d->nyloc);
  fprintf(FID, "%i\n", p2d->pcoords[0]);
  fprintf(FID, "%i\n", p2d->pcoords[1]);
  fprintf(FID, "%.16e\n", t);

  /* output the solution values and close the data set */
  for (j=0; j<p2d->nyloc; j++)
    for (i=0; i<p2d->nxloc; i++)
      fprintf(FID, "%.16e\n",u[idx(i,j,p2d->nxloc)]);
  fclose(FID);


  if (p2d->myid == 0) {
    /* now output a metadata file, containing general run information */
    FID = fopen("u_sol.txt","w");
    fprintf(FID, "%i\n", nx);
    fprintf(FID, "%i\n", ny);
    fprintf(FID, "%i\n", p2d->pdims[0]);
    fprintf(FID, "%i\n", p2d->pdims[1]);
    fprintf(FID, "%i\n", noutput);
    fclose(FID);
  }

} /* end output_mpi */
